upup/pkg/fi/cloudup/azuretasks/vmscaleset.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azuretasks
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-01/compute"
+	"github.com/Azure/go-autorest/autorest/to"
+	"k8s.io/klog/v2"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/azure"
+)
+
+// SubnetID contains the resource ID/names required to construct a subnet ID.
+type SubnetID struct {
+	SubscriptionID     string
+	ResourceGroupName  string
+	VirtualNetworkName string
+	SubnetName         string
+}
+
+// String returns the subnet ID in the path format.
+func (s *SubnetID) String() string {
+	return fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s/subnets/%s",
+		s.SubscriptionID,
+		s.ResourceGroupName,
+		s.VirtualNetworkName,
+		s.SubnetName)
+}
+
+// ParseSubnetID parses a given subnet ID string and returns a SubnetID.
+func ParseSubnetID(s string) (*SubnetID, error) {
+	l := strings.Split(s, "/")
+	if len(l) != 11 {
+		return nil, fmt.Errorf("malformed format of subnet ID: %s, %d", s, len(l))
+	}
+	return &SubnetID{
+		SubscriptionID:     l[2],
+		ResourceGroupName:  l[4],
+		VirtualNetworkName: l[8],
+		SubnetName:         l[10],
+	}, nil
+}
+
+//go:generate fitask -type=VMScaleSet
+
+// VMScaleSet is an Azure VM Scale Set.
+type VMScaleSet struct {
+	Name      *string
+	Lifecycle *fi.Lifecycle
+
+	ResourceGroup  *ResourceGroup
+	VirtualNetwork *VirtualNetwork
+	Subnet         *Subnet
+	StorageProfile *VMScaleSetStorageProfile
+	// RequirePublicIP is set to true when VMs require public IPs.
+	RequirePublicIP *bool
+	// SKUName specifies the SKU of of the VM Scale Set
+	SKUName *string
+	// Capacity specifies the number of virtual machines the VM Scale Set.
+	Capacity *int64
+	// ComputerNamePrefix is the prefix of each VM name of the form <prefix><base-36-instance-id>.
+	// See https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-instance-ids.
+	ComputerNamePrefix *string
+	// AdmnUser specifies the name of the administrative account.
+	AdminUser    *string
+	SSHPublicKey *string
+	// CustomData is the user data configuration
+	CustomData  *fi.ResourceHolder
```

Comment:
`fi.ResourceHolder` was removed in https://github.com/kubernetes/kops/pull/10472 and https://github.com/kubernetes/kops/pull/10478 so you should be able to replace it in a similar way that those PRs do.

Reply:
Sure, fixed!